### ISSUE TITLE
Found icon of unknown type 'unknown' in 'system/flatpak/flatpak/cc.nift.nsm/*', skipping it.

### ISSUE THREAD

**Issue by johanfleury** [flathub/cc.nift.nsm#3]
This Flatpak is triggering warnings in apt (and any other package manager that call appstreamcli) even without being installed:

```
$ flatpak list | grep 'cc\.nift\.nsm' # returns nothing
$

# appstreamcli refresh-cache --force

** (appstreamcli:18656): WARNING **: 11:20:37.242: Found icon of unknown type 'unknown' in 'system/flatpak/flatpak/cc.nift.nsm/*', skipping it.

** (appstreamcli:18656): WARNING **: 11:20:37.242: Found icon of unknown type 'unknown' in 'system/flatpak/flatpak/cc.nift.nsm/*', skipping it.
✔ Metadata cache was updated successfully.
```

**Comment by okias** [comment]
I also prefer to not be spammed by these warning for application I don't use.

**Comment by doczipeter** [comment]
Also mentioned @ https://github.com/flatpak/flatpak/issues/4642#issuecomment-1047623503

**Comment by smcv** [comment]
See also https://github.com/nifty-site-manager/nsm-flatpak/issues/1.

**Comment by ximion** [comment]
Fairly easy to fix:
```console
earth@sun:~$ appstreamcli validate cc.nift.nsm.appdata.xml 
W: cc.nift.nsm:~: console-app-no-binary
E: cc.nift.nsm:30: type-property-required icon ($APPID.png)
I: cc.nift.nsm:12: description-first-para-too-short
     Nift (aka nsm) is a cross-platform open source CLI website generator.
E: cc.nift.nsm:31: type-property-required icon ($APPID.png)

✘ Validation failed: errors: 2, warnings: 1, infos: 1, pedantic: 1
```
=> The `<icon/>` tags need a `type=` attribute - I guess the intent here was to use `stock` as icon type. See https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-icon
(also, you can pass `--explain` to `appstreamcli` to get a verbose explanation of the issues it found)

**Comment by mmieluch** [comment]
Ummm...

```
me@localhost:~$ appstreamcli validate cc.nift.nsm.appdata.xml --explain
File 'cc.nift.nsm.appdata.xml' does not exist.
✘ Validation failed:
```

That's the full output.

**Comment by mmieluch** [comment]
OK, force-refreshing cache worked for me (kudos to the OP):
```
appstreamcli refresh-cache --force
```

**Comment by laniakea64** [comment]
Hi, I'm still seeing these warnings even now, well after the linked pull request was merged and I've repeatedly run `flatpak update` and `sudo appstreamcli refresh --force` in the mean time.  Is there something else I need to do to purge some old cache or similar to stop these warnings?  Or is this just a matter of waiting for @n-ham to put the fixed version on Flathub and there's nothing to do about it on the user end yet?
(I don't have this app installed.)

Thanks

**Comment by hfiguiere** [comment]
For one thing, the desktop file has this: `Version=v2.4.12` and it's wrong.

https://specifications.freedesktop.org/desktop-entry-spec/latest/ar01s06.html

>  Version of the Desktop Entry Specification that the desktop entry conforms with. Entries that confirm with this version of the specification should use 1.5. Note that the version field is not required to be present. 

Second:

it's not installed in the right place:

https://github.com/nifty-site-manager/nsm-flatpak/blob/master/Makefile#L125
> 	`install -Dm644 cc.nift.nsm.desktop $(prefix)/$(datadir)/desktop/cc.nift.nsm.desktop`

It should be `applications`. `Not desktop`.

While I can't confirm whether this is the problem, it sure would be nice to fix.

**Comment by hfiguiere** [comment]
also
```
WARNING: Failed to validate desktop file /var/home/hub/source/flathub/cc.nift.nsm/repo/export/share/applications/cc.nift.nsm.desktop: /var/home/hub/source/flathub/cc.nift.nsm/repo/export/share/applications/cc.nift.nsm.desktop: error: value "ConsoleOnly;DeveloperTools;Productivity;Utilities;" for key "Categories" in group "Desktop Entry" contains an unregistered value "DeveloperTools"; values extending the format should start with "X-"
/var/home/hub/source/flathub/cc.nift.nsm/repo/export/share/applications/cc.nift.nsm.desktop: error: value "ConsoleOnly;DeveloperTools;Productivity;Utilities;" for key "Categories" in group "Desktop Entry" contains an unregistered value "Productivity"; values extending the format should start with "X-"
/var/home/hub/source/flathub/cc.nift.nsm/repo/export/share/applications/cc.nift.nsm.desktop: error: value "ConsoleOnly;DeveloperTools;Productivity;Utilities;" for key "Categories" in group "Desktop Entry" contains an unregistered value "Utilities"; values extending the format should start with "X-"
/var/home/hub/source/flathub/cc.nift.nsm/repo/export/share/applications/cc.nift.nsm.desktop: hint: value "ConsoleOnly;DeveloperTools;Productivity;Utilities;" for key "Categories" in group "Desktop Entry" does not contain a registered main category; application might only show up in a "catch-all" section of the application menu
```

**Comment by hfiguiere** [comment]
and

https://github.com/nifty-site-manager/nsm-flatpak/blob/master/cc.nift.nsm.appdata.xml#L30

These are invalid too.

**Comment by hfiguiere** [comment]
it is published.
but the page will be missing still since this is a CLI flatpak.

**Comment by ximion** [comment]
> and
> 
> https://github.com/nifty-site-manager/nsm-flatpak/blob/master/cc.nift.nsm.appdata.xml#L30
> 
> These are invalid too.

Nah, those are fine, the placeholder gets replaced during build and it is strongly recommended that `console-application` apps have an icon set.

**Comment by ximion** [comment]
> [...] Or is this just a matter of waiting for @n-ham to put the fixed version on Flathub and there's nothing to do about it on the user end yet? [...]

Exactly this is the case! Once a new version of the app is on Flathub, the issue should be gone.

**Comment by hfiguiere** [comment]
> Nah, those are fine, the placeholder gets replaced during build and it is strongly recommended that `console-application` apps have an icon set.

they didn't get replaced with the current version. I checked.

**Comment by ximion** [comment]
> > Nah, those are fine, the placeholder gets replaced during build and it is strongly recommended that `console-application` apps have an icon set.
> 
> they didn't get replaced with the current version. I checked.

Brr... Then they should simply be `cc.nift.nsm` (the placeholder doesn't make much sense anyway), but icons should be present in the metainfo files for console apps.

**Comment by hfiguiere** [comment]
it has a .desktop file so it use that to find the icons.

**Comment by ximion** [comment]
> it has a .desktop file so it use that to find the icons.

No, it doesn't. Believe me, I wrote the code for this ;-)
The desktop search will only ever happen for `desktop-application` components types which also have a `launchable` entry of type `desktop-id` or where some heuristics on their component-ID apply. It is never run for any other component type.
The specification even documents it like this (the behavior is only mentioned for the desktop-app type and omitted from any other component type, and even explicitly called out for e.g. the `name` tag: https://www.freedesktop.org/software/appstream/docs/chap-CollectionData.html#tag-ct-name).

**Comment by GalaxyCurse** [comment]
And it's back... 
`(17/28) Updating the appstream cache...

** (appstreamcli:15171): WARNING **: 10:11:11.060: Found icon of unknown type 'unknown' in 'system/flatpak/flatpak/cc.nift.nsm/*', skipping it.

** (appstreamcli:15171): WARNING **: 10:11:11.060: Found icon of unknown type 'unknown' in 'system/flatpak/flatpak/cc.nift.nsm/*', skipping it.
✔ Metadata cache was updated successfully.
`
Arch Linux October 24th, 2022

**Comment by n-ham** [comment]
Gah! Sorry anyone/everyone affected by that! I was planning to smash out a v3.0 of Nift with a bunch of improvements and then hit up @ximion when I was ready to release to make sure we had all this cleaned up (funnily I had indeed been speaking to someone with a very similar name to ximion when I first started playing with the json parsing). 

I've been busily getting ready to do so, but was first working on my own json parser to use (not the biggest fan or rapidjson or simdjson) which has blown out to putting my own hat in the ring for ways to enforce safety with c++ projects (which coincidentally will later involve rewriting the template/scripting languages in Nift and using them for templating with c++ projects).

I'm still active at the moment though so should be able to get this sorted sooner rather than later. If anyone happens to easily know what the actual issue is that'd save me having to go exploring to figure things out when I have very limited knowledge of flathub/pak.
